src/goLanguageServer.ts

```diff
@@ -88,18 +94,38 @@ export async function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
 			},
 			revealOutputChannelOn: RevealOutputChannelOn.Never,
 			middleware: {
-				provideDocumentFormattingEdits: (document: vscode.TextDocument, options: FormattingOptions, token: vscode.CancellationToken, next: ProvideDocumentFormattingEditsSignature) => {
-					if (!config.features.format) {
-						return [];
+				didSave: (data: vscode.TextDocument, next: NextSignature<vscode.TextDocument, void>) => {
+					// If the user only wants diagnostics on save, show the diagnostics now.
+					if (config.features.diagnostics == DiagnosticFrequency.OnSave) {
+						const diagnostics = latestDiagnostics.get(data.uri.toString());
+						c.diagnostics.set(data.uri, diagnostics);
 					}
-					return next(document, options, token);
+					return next(data, null);
 				},
 				handleDiagnostics: (uri: vscode.Uri, diagnostics: vscode.Diagnostic[], next: HandleDiagnosticsSignature) => {
```

Comment:
I am not familiar as to when exactly is `handleDiagnostics` called? Is it for every file change or file open?

Reply:
It's called whenever a diagnostics notification comes in - for `gopls`, we send diagnostics when needed for any file event (open, change, close, save, create, delete).

Reply:
In that case, isn't it better to use the "gopls" settings such that `gopls` avoids calculating the diagnostics on file change if the user wants them only on file save?

Reply:
`gopls` always parses and type checks on file changes, as that information is necessary for other `gopls` features, so there will be no benefits to pushing this configuration into `gopls` itself. That's the reasoning for this PR - to avoid displaying diagnostics if users don't want to see them without affecting `gopls`.

Reply:
understood